src/main/java/teammates/ui/newcontroller/Action.java

```diff
@@ -83,6 +85,13 @@ private void initAuthInfo() {
 
         userInfo = gateKeeper.getCurrentUser();
         authType = userInfo == null ? AuthType.PUBLIC : AuthType.LOGGED_IN;
+
+        if (userInfo != null && userInfo.isAdmin && req.getParameter(Const.ParamsNames.USER_ID) != null
+                && HttpGet.METHOD_NAME.equals(req.getMethod())) {
+            // Only allow masquerade on non-modifying operations to prevent accidental change to user data
```

Comment:
But most of our test depend on Masquerade mode to simulate user actions including adding/modifying action.

Reply:
I forgot about the E2E tests here. Can't say I like it, but for operational reasons, I guess it needs to stay that way.